### PR TITLE
feat: add cache interface and implementation for redis,inmemory

### PR DESCRIPTION
**Pull request by srikanthccv**
Part of https://github.com/SigNoz/signoz/issues/1988

This adds a cache interface that can be used for general purposes. It currently supports redis and memory as a backend storage mechanism. 

I will be sending series of PRs for the #1988 to make it easier to review

### CODE REVIEW

**Comment by srikanthccv** [comment]
This is updated to use `v3.QueryRangeParamsV3` and is ready for reviews.

**Comment by srikanthccv** [comment]
It has been a while since marking this ready for review. Please review. Or let me know if there was something I could do to get reviews faster.

File: pkg/query-service/cache/inmemory/options.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package inmemory
+
+import (
+	"time"
+)
+
+const (
+	defaultTTL             = 5 * time.Minute
```
Comment (nityanandagohain) [inline pkg/query-service/cache/inmemory/options.go:8]: Shouldn't default ttl be forever with no cleanup required?

Comment (srikanthccv) [reply pkg/query-service/cache/inmemory/options.go:8]: I was thinking not to let this grow forever and cause memory issues, but I see where you are coming from. If I understand you correctly, the default should be set to the maximum possible, but the user who creates the cache instances should pass what's appropriate for them, is that correct?

Comment (nityanandagohain) [reply pkg/query-service/cache/inmemory/options.go:8]: Yeah, that's where I was coming form. Though any serious user will set the values explicitly.

Comment (srikanthccv) [reply pkg/query-service/cache/inmemory/options.go]: Updated

**Review by nityanandagohain — APPROVED** [review]

File: pkg/query-service/cache/inmemory/cache.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// Close does nothing
+func (c *cache) Close() error {
+	return nil
+}
+
+// Configuration returns the cache configuration
+func (c *cache) Configuration() *Options {
+	return nil
```
Comment (vikrantgupta25) [inline pkg/query-service/cache/inmemory/cache.go:72]: should have returned the options from here for the configuration

**Pull request merged**